src/components/ActionManager/ActionManager.js

```diff
@@ -0,0 +1,119 @@
+import Cosmos from "@lunie/cosmos-js"
+import config from "src/config"
+import { getSigner } from "./signer"
+import transaction from "./transactionTypes"
+
+import { formatCosmosArguments, convertCurrencyData } from "./cosmosMessages"
+
+class ActionManager {
+  setContext(context) {
+    if (!context) {
+      return
+    }
+    this.context = context
+    this.cosmos = new Cosmos(this.context.url || "", this.context.chainId || "")
+  }
+
+  async simulate(type, transactionProperties) {
+    if (!this.context) {
+      throw Error(`This modal has no context.`)
+    }
+
+    if (!this.context.connected) {
+      throw Error(
+        `Currently not connected to a secure node. Please try again when Lunie has secured a connection.`
+      )
+    }
+
+    const txArguments = formatCosmosArguments(
+      this.context,
+      type,
+      transactionProperties,
+      true
+    )
+
+    if (txArguments === null) {
+      throw Error(
+        `Did not complete transaction simulation. Invalid message type: ${type}.`
+      )
+    }
+
+    const message = this.createCosmosMessage(
+      type,
+      this.context.userAddress,
+      txArguments
+    )
+    const gasEstimate = await message.simulate({
+      memo: transactionProperties.memo
+    })
+    return gasEstimate
+  }
+
+  async send(type, transactionProperties, txMetaData) {
+    if (!this.context) {
+      throw Error(`This modal has no context.`)
+    }
+
+    if (!this.context.connected) {
+      throw Error(
+        `Currently not connected to a secure node. Please try again when Lunie has secured a connection.`
+      )
+    }
+
+    const txArguments = formatCosmosArguments(
+      this.context,
+      type,
+      transactionProperties
+    )
+
+    if (txArguments === null) {
+      throw Error(
+        `Did not complete transaction. Invalid message type: ${type}.`
+      )
+    }
+
+    const { gasEstimate, gasPrice, submitType, password } = txMetaData
+
+    const localKeyPairName = this.context.localKeyPairName
+    const signer = getSigner(config, submitType, { localKeyPairName, password })
+
+    let message
+    if (type === transaction.WITHDRAW) {
+      message = this.createMultiMessage(
+        type,
+        this.context.userAddress,
+        txArguments
+      )
+    } else {
+      message = this.createCosmosMessage(
+        type,
+        this.context.userAddress,
+        txArguments
+      )
+    }
+
+    const { included } = await message.send(
+      {
+        gas: String(gasEstimate),
+        gas_prices: convertCurrencyData([gasPrice]),
+        memo: transactionProperties.memo
+      },
+      signer
+    )
+    await included()
+  }
+
+  createCosmosMessage(type, senderAddress, txArguments) {
+    return this.cosmos[type](senderAddress, txArguments)
```

Comment:
if this is not self explanatory, and needs a `createCosmosMessage` should we change the api of `cosmos-js`?

Reply:
It was probably to keep it synchronised with `createMultiMessage`, and `cosmos[type]` isn't instantly obvious what it returns. If you're up for changing the API we can work on that.

Reply:
Always up for improving the API. cosmos-js is our library so we should try to not hack around it but improve it so it fits our needs.